go/cmd/data/cmd/containerwatcher/kubeapi_watcher.go

```diff
@@ -0,0 +1,85 @@
+package containerwatcher
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/lyft/flytestdlib/logger"
+	v13 "k8s.io/api/core/v1"
+	v12 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+type ContainerInformation struct {
+	Namespace string
+	PodName   string
+	Name      string
+}
+
+type kubeAPIWatcher struct {
+	coreClient v1.CoreV1Interface
+	info       ContainerInformation
+}
+
+func (k kubeAPIWatcher) wait(ctx context.Context, info ContainerInformation, f func(ctx context.Context, pod *v13.Pod) (bool, error)) error {
+	s := fields.OneTermEqualSelector("metadata.name", info.PodName)
+	watcher, err := k.coreClient.Pods(info.Namespace).Watch(v12.ListOptions{
+		Watch:         true,
+		FieldSelector: s.String(),
+	})
+	if err != nil {
+		return err
+	}
+	defer watcher.Stop()
+	for {
+		select {
+		case v := <-watcher.ResultChan():
+			p := v.Object.(*v13.Pod)
+			if stop, err := f(ctx, p); err != nil {
+				return err
+			} else if stop {
+				return nil
+			}
+		case <-ctx.Done():
+			logger.Infof(ctx, "Pod [%s/%s] watcher canceled", info.Namespace, info.PodName)
+			return ErrTimeout
+		}
+	}
+}
+
+func (k kubeAPIWatcher) WaitToStart(ctx context.Context) error {
+	f := func(ctx context.Context, p *v13.Pod) (bool, error) {
+		for _, c := range p.Status.ContainerStatuses {
+			if c.Name == k.info.Name {
+				if c.Ready {
```

Comment:
hmm I guess you are saying that in case we are pending in imagepull or we have failed then we should exit too, just with a failure exit code. Suggest If I should just remove this option for now?

Reply:
I am actually not using this mode at all right now.